binding.cc

```diff
@@ -3178,754 +3352,824 @@ typedef struct sn_crypto_stream_salsa20_xor_state {
+  auto *next_block = state_data->next_block;
 
-  unsigned char *next_block = state->next_block;
+  size_t m_size = m.size_bytes();
+  auto *c_ptr = c.data();
+  auto *m_ptr = m.data();
 
-  if (state->remainder) {
+  if (state_data->remainder) {
     uint64_t offset = 0;
-    int rem = state->remainder;
+    int rem = state_data->remainder;
 
     while (rem < 64 && offset < m_size) {
-      c[offset] = next_block[rem]  ^ m[offset];
-      offset++;
-      rem++;
+      c_ptr[offset] = next_block[rem] ^ m_ptr[offset];
+      ++offset;
+      ++rem;
     }
 
-    c += offset;
-    m += offset;
+    c_ptr += offset;
+    m_ptr += offset;
     m_size -= offset;
-    state->remainder = rem == 64 ? 0 : rem;
+    state_data->remainder = (rem == 64) ? 0 : rem;
 
-    if (!m_size) return NULL;
+    if (m_size == 0) return;
   }
 
-  state->remainder = m_size & 63;
-  m_size -= state->remainder;
-  crypto_stream_salsa20_xor_ic(c, m, m_size, state->n, state->block_counter, state->k);
-  state->block_counter += m_size / 64;
-
-  if (state->remainder) {
-    sodium_memzero(next_block + state->remainder, 64 - state->remainder);
-    memcpy(next_block, m + m_size, state->remainder);
+  state_data->remainder = m_size & 63;
+  size_t main_len = m_size - state_data->remainder;
 
-    crypto_stream_salsa20_xor_ic(next_block, next_block, 64, state->n, state->block_counter, state->k);
-    memcpy(c + m_size, next_block, state->remainder);
-
-    state->block_counter++;
-  }
+  crypto_stream_salsa20_xor_ic(
+    c_ptr, m_ptr, main_len, state_data->n, state_data->block_counter, state_data->k
+  );
 
-  return NULL;
-}
+  state_data->block_counter += main_len / 64;
 
-js_value_t *
-sn_crypto_stream_salsa20_xor_wrap_final (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, crypto_stream_salsa20_xor_wrap_final)
+  if (state_data->remainder) {
+    sodium_memzero(next_block + state_data->remainder, 64 - state_data->remainder);
+    memcpy(next_block, m_ptr + main_len, state_data->remainder);
 
-  SN_ARGV_BUFFER_CAST(sn_crypto_stream_salsa20_xor_state *, state, 0)
+    crypto_stream_salsa20_xor_ic(
+      next_block, next_block, 64, state_data->n, state_data->block_counter, state_data->k
+    );
+    memcpy(c_ptr + main_len, next_block, state_data->remainder);
 
-  SN_THROWS(state_size != sizeof(sn_crypto_stream_salsa20_xor_state), "state must be 'crypto_stream_salsa20_xor_STATEBYTES' bytes")
+    state_data->block_counter++;
+  }
+}
 
-  sodium_memzero(state->n, sizeof(state->n));
-  sodium_memzero(state->k, sizeof(state->k));
-  sodium_memzero(state->next_block, sizeof(state->next_block));
-  state->remainder = 0;
+static inline void
+sn_crypto_stream_salsa20_xor_wrap_final(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state
+) {
+  assert(state.size_bytes() == sizeof(sn_crypto_stream_salsa20_xor_state));
+  auto state_data = reinterpret_cast<sn_crypto_stream_salsa20_xor_state *>(state.data());
```

Comment:
This could be improved by `libjstl` adding `js_typedarray_span_of_t<T>` similar to `js_arraybuffer_span_of_t<T>`.